### PR TITLE
fix(frontend): crashes on fn ptr

### PR DESCRIPTION
**Pull request by W95Psp**
This commit fixes the issue #757: the frontend was assuming functions expressions to be of a very specific shape. This commit fixes that.

### CODE REVIEW

File: frontend/exporter/src/types/copied.rs [hunk 1]
```diff
@@ -2412,66 +2412,50 @@ pub enum ExprKind {
...
+                    id: def_id.sinto(gstate)
+                });
+                let mut translated_generics = generics.sinto(gstate);
+                let tcx = gstate.base().tcx;
+                r#trait = (|| {
+                    let assoc_item = tcx.opt_associated_item(*def_id)?;
+                    let assoc_trait = tcx.trait_of_item(assoc_item.def_id)?;
+                    let trait_ref = ty::TraitRef::new(tcx, assoc_trait, generics.iter());
+                    let impl_expr = {
+                        // TODO: we should not wrap into a dummy binder
```
Comment (franziskuskiefer) [inline frontend/exporter/src/types/copied.rs:2430]: Should this todo be addressed?

**Review by franziskuskiefer — APPROVED** [review]
Thanks.

It looks like the following two functions get translated very differently. Is that on purpose?

```rust
fn do_twice(f: fn(i32) -> i32, arg: i32) -> i32 {
    f(arg) + f(arg)
}

fn do_again<F>(f: F, arg: i32) -> i32
where
    F: Fn(i32) -> i32,
{
    f(arg) + f(arg)
}
```

```ocaml
let do_again
      (#v_F: Type0)
      (#[FStar.Tactics.Typeclasses.tcresolve ()] i1: Core.Ops.Function.t_Fn v_F i32)
      (f: v_F)
      (arg: i32)
    : i32 =
  (Core.Ops.Function.f_call #v_F #i32 f (arg <: i32) <: i32) +!
  (Core.Ops.Function.f_call #v_F #i32 f (arg <: i32) <: i32)

let do_twice (f: (i32 -> i32)) (arg: i32) : i32 = f arg +! f arg
```

File: frontend/exporter/src/types/copied.rs (hunk 1 above)
Comment (W95Psp) [reply frontend/exporter/src/types/copied.rs:2430]: @Nadrieril and I are looking at those things in parallel!

**Comment by W95Psp** [comment]
Yes, higher-order functions are a bit problematic, that's related to what @paulmure says in https://github.com/hacspec/hax/issues/549#issuecomment-2218838180

**Pull request merged**